jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -0,0 +1,621 @@
+             * uncompressed output chunks.</p>
+             * <p>Implementations should return an {@link Chunk.Error error chunk} in case
+             * of transformation errors.</p>
+             * <p>Exceptions thrown by this method are equivalent to returning an error chunk.</p>
+             * <p>Implementations of this method must arrange to {@link Chunk#release() release}
+             * the input chunk, unless they return it as is.
+             * The output chunk is released by the code that uses this Transformer.</p>
+             *
+             * @param rawChunk the input chunk to transform
+             * @return the transformed output chunk
+             */
+            @Override
+            protected abstract Chunk transform(Chunk rawChunk);
+        }
+    }
+
+    /**
+     * <p>A content sink that writes the content to its implementation (a socket, a file, etc.).</p>
+     */
+    public interface Sink
+    {
+        /**
+         * <p>Wraps the given content sink with an {@link OutputStream}.</p>
+         *
+         * @param sink the sink to write to
+         * @return an OutputStream that writes to the content sink
+         */
+        public static OutputStream asOutputStream(Sink sink)
+        {
+            return new ContentSinkOutputStream(sink);
+        }
+
+        /**
+         * <p>Wraps the given content sink with a {@link Flow.Subscriber}.</p>
+         *
+         * @param sink the sink to write to
+         * @return a Subscriber that writes to the content sink
+         */
+        public static Flow.Subscriber<Chunk> asSubscriber(Sink sink)
+        {
+            return new ContentSinkSubscriber(sink);
+        }
+
+        /**
+         * <p>Blocking version of {@link #write(boolean, Callback, ByteBuffer...)}.</p>
+         *
+         * @param sink the sink to write to
+         * @param last whether the ByteBuffers are the last to write
+         * @param buffers the ByteBuffers to write
+         * @throws IOException if the write operation fails
+         */
+        public static void write(Sink sink, boolean last, ByteBuffer... buffers) throws IOException
+        {
+            try (Blocking.Callback callback = Blocking.callback())
+            {
+                sink.write(last, callback, buffers);
+                callback.block();
+            }
+        }
+
+        /**
+         * <p>Writes the given {@link Chunk}, notifying the {@link Callback}
+         * when the write is complete.</p>
+         *
+         * @param chunk the chunk to write
+         * @param callback the callback to notify when the write operation is complete
+         * @see #write(boolean, Callback, ByteBuffer...)
+         */
+        public default void write(Chunk chunk, Callback callback)
+        {
+            if (chunk instanceof Content.Chunk.Error error)
+                callback.failed(error.getCause());
+            else
+                write(chunk.isLast(), callback, chunk.getByteBuffer());
```

Comment:
This is actually wrong because it ignores the release conversation. Needs to be:
```suggestion
            if (chunk instanceof Content.Chunk.Error error)
            {
                chunk.release();
                callback.failed(error.getCause());
            }
            else
                write(chunk.isLast(), Callback.from(chunk::release, callback), chunk.getByteBuffer());
```

Reply:
For the happy path, that's _another_ allocation on the fast path.
For the error path, since that `release()` call is a no-op I wonder if we should not document it as such on `Error`.